### DISCUSSION TITLE
How to set LabelStyle to display non overlapping.

### DISCUSSION THREAD

**Discussion by lucinelouhi** [Mapsui/Mapsui#2617] [Q&A]
How to set LabelStyle to display non overlapping.
![QQ图片20240504173738](https://github.com/Mapsui/Mapsui/assets/166964682/de352854-872f-4c0b-a358-918deeda3d38)。
How to set LabelStyle to display non overlapping. This image has overlapping labels, how can the number of displayed labels be different at different scales, without conflict or coverage. Thank you for overlapping like this to make it clearer to see.

**Comment by charlenni** [comment]
That is not possible in the moment. For this you need to know, how large the label is. If you know this, you could create your own layer, that is capable of deciding, which feature to draw or not.

As side note: this isn't an easy problem. Normally there is a symbol and a label for a feature. Then there are different ways to display both of them (position relative to each other) or only one, if there isn't enough space or none of them.

**Reply by lucinelouhi** [reply]
Indeed, if necessary, it is necessary to perform spatial topology operations on the rectangular range of the drawn Label. Another way is to display it at the current scale. But it's up to everyone to decide which ones are displayed and which ones are not. It is estimated that it is necessary to modify the underlying drawing algorithm or customize the Style to make a decision. Additionally, I tested Mapsui's performance in WinForm and it was okay. But the feeling of font drawing is not particularly beautiful.

**Reply by charlenni** [reply]
Yes, no one made a Layer for this. Perhaps you want to contribute this?

What do you mean with handling fonts?

**Reply by lucinelouhi** [reply]
The font for the Chinese annotation layer is not very aesthetically pleasing. I'm not sure if this is a problem that needs to be adjusted. As for the issue with that layer, I will see if I can resolve it by referring to other code.

**Reply by lucinelouhi** [reply]
The MeteoInfo_csharp library seems to have solved this problem. But I haven't found similar code there yet.

**Reply by lucinelouhi** [reply]
It should be this MeteoInfoLib_CSharp project.

**Reply by charlenni** [reply]
Do you have an image? Do you use a font, that is capable of displaying this characters?

Mapsui use Harfbuzz and RichTextKit to render text. Both should be capable of displaying any characters.
